File: rpi1/rpi1_heartbeat.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <mosquitto.h>

static volatile sig_atomic_t running = 1;

static void handle_signal(int sig)
{
    (void)sig;
    running = 0;
}

int main(void)
{
    struct mosquitto *mosq;
    const char *topic = "parking/rpi1/heartbeat";
    const char *payload = "{\"node\":\"rpi1\",\"status\":\"online\"}";

    signal(SIGINT, handle_signal);
    signal(SIGTERM, handle_signal);

    mosquitto_lib_init();
    mosq = mosquitto_new(NULL, true, NULL);
    if (!mosq) {
        mosquitto_lib_cleanup();
        return 1;
    }

    if (mosquitto_connect(mosq, "192.168.4.1", 1883, 60) != MOSQ_ERR_SUCCESS) {
        mosquitto_destroy(mosq);
        mosquitto_lib_cleanup();
        return 1;
    }

    while (running) {
        mosquitto_publish(mosq, NULL, topic,
                          strlen(payload), payload, 1, false);
        sleep(2);
    }

    mosquitto_disconnect(mosq);
    mosquitto_destroy(mosq);
    mosquitto_lib_cleanup();

    return 0;
}

File: rpi1/rpi1_heartbeat.py
import json
import os
import signal
import time

import paho.mqtt.client as mqtt


BROKER_HOST = os.getenv("MQTT_HOST", "192.168.4.1")
BROKER_PORT = int(os.getenv("MQTT_PORT", "1883"))
TOPIC = "parking/rpi1/heartbeat"
PAYLOAD = {"node": "rpi1", "status": "online"}
INTERVAL_SECONDS = 2

running = True


def handle_signal(signum, frame):
    global running
    running = False


def main():
    signal.signal(signal.SIGINT, handle_signal)
    signal.signal(signal.SIGTERM, handle_signal)

    client = mqtt.Client()
    client.connect(BROKER_HOST, BROKER_PORT, 60)
    client.loop_start()

    try:
        while running:
            client.publish(TOPIC, json.dumps(PAYLOAD), qos=1)
            time.sleep(INTERVAL_SECONDS)
    finally:
        client.loop_stop()
        client.disconnect()


if __name__ == "__main__":
    main()
